mopidy_musicbox_webclient/static/js/images.js

```diff
@@ -4,105 +4,279 @@
+         *                        used in the last.fm lookups if Mopidy does not provide the image
+         *                        directly. Can be one of 'small', 'medium', 'large',
+         *                        'extralarge' (default), or 'mega'.
+         */
+        setArtistImage: function (artist_uri, track_uri, img_element, mopidy, size) {
+            // Set default immediately while we're busy retrieving actual image.
+            $(img_element).attr('src', images.DEFAULT_ARTIST_URL)
+            if (!artist_uri && !track_uri) {
+                return
+            }
+            size = size || 'extralarge'
+
+            if (artist_uri) {
+                // Use artist as starting point for retrieving image.
+                mopidy.library.getImages({'uris': [artist_uri]}).then(function (imageResults) {
+                    var uri = Object.keys(imageResults)[0]
+                    if (imageResults[uri].length > 0) {
+                        $(img_element).attr('src', imageResults[uri][0].uri)
+                    } else {
+                        // Fall back to using track as starting point for retrieving image.
+                        images._setArtistImageFromTrack(track_uri, img_element, mopidy, size)
+                    }
+                })
+            }
+        },
+
+        /* Set artist image using the supplied Mopidy track URI. */
+        _setArtistImageFromTrack: function (uri, img_element, mopidy, size) {
+            mopidy.library.lookup({'uris': [uri]}).then(function (resultDict) {
+                var uri = Object.keys(resultDict)[0]
+                var track = resultDict[uri][0]
+                var artist = images._getArtistInfo(track)
+
+                if (artist.mbid) {
+                    images._setLastFmArtistImage(artist.mbid, img_element, size)
+                } else {
+                    // Look up unique MusicBrainz ID for artist first using the available track information
+                    images._getLastFmArtistInfo(track).then(function (artist) {
+                        images._setLastFmArtistImage(artist.mbid, img_element, size)
+                    }, function (code, message) {
+                        $(img_element).attr('src', images.DEFAULT_ARTIST_URL)
+                        console.error('Error retrieving artist info from last.fm. (%s: %s)', code, message)
+                    })
+                }
+            })
         },
 
-        getArtistImage: function (artist, images, size) {
-            var defUrl = 'images/user_24x32.png'
-            $(images).attr('src', defUrl)
-            if (!artist || artist.length === 0) {
+        /* Set artist image using the supplied artist MusicBrainz ID. */
+        _setLastFmArtistImage: function (mbid, img_element, size) {
+            if (!mbid) {
+                // Avoid expensive last.fm call if tag information is missing.
+                $(img_element).attr('src', images.DEFAULT_ALBUM_URL)
```

Comment:
Absolutely. Isn't it impressive how I managed to get this mixed up, and then write a test that passes as well?